File: ClassHierarchyMaintainability/Object.r
#ifndef HIERARCHY_OBJECT_R
#define HIERARCHY_OBJECT_R

#include <stdio.h>
#include <stdarg.h>

struct Object 
{
	struct Class * clas;
};

struct Class 
{
	const struct Object _;
	char * name;
	const struct Class * superClass;
	void *(*constructor)(void *self, va_list *list);
	void *(*destructor)(void *self);
	int (*differ)(const void *self, const void *other);
	void (*puto)(const void *self, FILE *fp);	
};

#endif

File: ClassHierarchyMaintainability/Object.c
#include <stdio.h>
#include <stdlib.h>
#include "Object.r"

void * new(const void *_self, ...)
{
	return malloc(sizeof(struct Object));
}

static void * Object_constructor(void *_self, va_list *list)
{
	return _self;
}

static void * Object_destructor(void *_self, va_list *list)
{
	return _self;
}

static int Object_differ(const void *_self, const void *other)
{
	return (_self == other);
}

static int Object_puto(const void *_self, FILE *fp)
{
	const struct Class * clas = classOf(_self);
	return fprintf(fp, "\nObject:\n\tName: %s\n\tMemory: %p", clas->name, _self);
}

